build-logic/basics/src/main/kotlin/gradlebuild/basics/BuildEnvironment.kt

```diff
@@ -20,14 +20,28 @@ import gradlebuild.basics.BuildParams.CI_ENVIRONMENT_VARIABLE
 import org.gradle.api.JavaVersion
 import org.gradle.api.Project
 import org.gradle.api.file.Directory
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.Property
 import org.gradle.api.provider.Provider
 import org.gradle.internal.os.OperatingSystem
+import org.gradle.kotlin.dsl.*
 
 
-fun Project.repoRoot() = layout.projectDirectory.parentOrRoot()
+abstract class BuildEnvironmentExtension(val project: Project) {
+    abstract val gitCommitId: Property<String>
+    abstract val gitBranch: Property<String>
+    abstract val repoRoot: DirectoryProperty
+}
 
 
+// `generatePrecompiledScriptPluginAccessors` task invokes this method without `gradle.build-environment` applied
 private
+fun Project.getBuildEnvironmentExtension(): BuildEnvironmentExtension? = rootProject.extensions.findByType(BuildEnvironmentExtension::class.java)
```

Comment:
You can detect whether this runs within `generatePrecompiledScriptPluginAccessors` by checking the project name. Maybe that would be simpler and you could provider a dummy extension here in that case.